pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,922 @@
+	}
+
+	ensureNameOrGenerateName(token)
+	// we check the Name directly. because of the ensure... we know that
+	// GenerateName is not set. as it squashes the name in that case.
+	if token.ObjectMeta.Name != "" {
+		// reject creation of a token which already exists
+		currentSecret, err := t.secretClient.Get(TokenNamespace, token.Name, metav1.GetOptions{})
+		if err == nil && currentSecret != nil {
+			return nil, apierrors.NewAlreadyExists(group, token.Name)
+		}
+	}
+
+	// reject user-provided token value, or hash
+	if token.Status.TokenValue != "" {
+		return nil, apierrors.NewBadRequest("User provided token value is not permitted")
+	}
+	if token.Status.TokenHash != "" {
+		return nil, apierrors.NewBadRequest("User provided token hash is not permitted")
+	}
+
+	// Get derived User information for the token.
+	//
+	// NOTE: The creation process is not given `AuthProvider`, `User-` and `GroupPrincipals`.
+	// ..... This information have to be retrieved from somewhere else in the system.
+	// ..... This is in contrast to the Norman tokens who get this information either
+	// ..... as part of the Login process, or by copying the information out of the
+	// ..... base token the new one is derived from. None of that is possible here.
+	//
+	// A User's `AuthProvider` information is generally captured in their associated
+	// `UserAttribute` resource. This is what we retrieve and use here now to fill these fields
+	// of the token to be.
+	//
+	// `ProviderInfo` is not supported. Norman tokens have it as legacy fallback to hold the
+	// `access_token` data managed by OIDC-based auth providers. The actual primary storage for
+	// this is actually a regular k8s Secret associated with the User.
+	//
+	// `UserPrincipal` is filled in part with standard information, and in part from the
+	// associated `User`s fields.
+
+	attribs, err := t.userAttributeClient.Get(token.Spec.UserID, metav1.GetOptions{})
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve user attributes of %s: %w",
+			token.Spec.UserID, err))
+	}
+	if attribs == nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to get user attributes of %s",
+			token.Spec.UserID))
+	}
+
+	if len(attribs.ExtraByProvider) != 1 {
+		return nil, apierrors.NewInternalError(fmt.Errorf("bad user attributes: bogus ExtraByProvider, empty or ambigous"))
+	}
+
+	user, err := t.userClient.Get(token.Spec.UserID, metav1.GetOptions{})
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve user %s: %w",
+			token.Spec.UserID, err))
+	}
+	if user == nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to get user %s", token.Spec.UserID))
+	}
+
+	// Generate secret and its hash
+
+	tokenValue, hashedValue, err := t.hasher.MakeAndHashSecret()
+	if err != nil {
+		return nil, err
+	}
+
+	// (len == 1) => The single key in the UA map names the auth provider
+	// (and where to look in GPs, if we were using GPs)
```

Comment:
Here and everywhere else: we should avoid using non-ubiquitous abbreviations in comments and identifiers (var, func, package etc names), but especially in comments. It's absolutely unclear what UA and GP means in this comment.